lib/Map/RegionProvider.js

```diff
@@ -137,58 +149,86 @@ RegionProvider.prototype.setDisambigProperties = function(dp) {
+    // get the list of IDs that we will attempt to match against for this column
+    var url = baseuri.setQuery('valueReference', propName).toString();
+
+    if (corsProxy.shouldUseProxy(url)) {
+        url = corsProxy.getURL(url);
     }
-    for (var i = 0; i < obj.member.length; i++) {
-        if (this.regions[i] === undefined) {
-            this.regions[i] = {};
+    return loadText(url)    
+        .then(function(xml) {
+            var obj = xml2json(xml);
+
+            if (!defined(obj.member)) {
+                console.log(xml);
+                var exception = defined(obj.Exception) ? ('<br/><br/>' + obj.Exception.ExceptionText) : '';
+                throw new ModelError({ title: 'CSV region mapping', message: 'Couldn\'t load region boundaries for region ' + propName + exception});
+            }
+
+            if (!(obj.member instanceof Array)) {
+                obj.member = [obj.member];
+            }
+            if (obj.member.length === 0) {
+                throw new ModelError({ title: 'CSV region mapping', message: 'Zero region boundaries found for region ' + propName });
+            }
+            return { values: obj.member.map(function(m) { return m[propName] }) };
+        });
+};
+
+/** 
+ *  Given a list of region IDs in feature ID order, apply server replacements if needed, and build an index.
+ */
+RegionProvider.prototype.processRegionIds = function(values, propertyName, replacementsProp) {
+    var that = this;
+    var isNumeric = (typeof values[0] === 'number');
+    // There is also generally a `layer` and `property` property in this file, which we ignore for now.
+    return values.map(function(id, i) {
+
+        if (that.regions[i] === undefined) {
+            that.regions[i] = {};
+        }
+        
+        if (!isNumeric) {
+            id = id.toLowerCase();
+            // we apply server-side replacements while loading. If it ever turns out we need
+            // to store the un-regexed version, we should add a line here.
+            id = that.applyReplacements(id, replacementsProp);
         }
-        var prop = (!propertyName ? obj.member[i][this.regionProp] : obj.member[i][propertyName]);
-        // we apply server-side replacements while loading. If it ever turns out we need
-        // to store the un-regexed version, we should add a line here.
-        // Potentially we could also convert numeric IDs to integers for faster matching.
-        prop = this.applyReplacements(prop.toLowerCase(), replacementsProp);
 
         // either { id: '123' } or { id: 'Bunbury', STE_NAME: 'VIC' }
         if (!propertyName) {
-            this.regions[i].id = prop;
+            that.regions[i].id = id;
         } else {
-            this.regions[i][propertyName] = prop;
+            that.regions[i][propertyName] = id;
         }
 
         // store a lookup by attribute, for performance.
-        this._idIndex[prop] = i;
+        that._idIndex[id] = i;
```

Comment:
Interesting point! As the code stands, this is never an issue, because `_idIndex` isn't even consulted when disambiguation is in question (see findRegionIndex). By definition, this kind of a hash table isn't much use when you have ambiguous values.

But you've prompted me to finish the job of the index, so it's used for disambiguation, too. (The disambiguation property itself doesn't get an index, because it wouldn't be useful.)